### PR TITLE
People screen ui

### PR DESCRIPTION
**Pull request by ylecollen**
Implement the new sticky headers feature

### CODE REVIEW

**Comment by ylecollen** [comment]
Review after testing it
-> The people screen flickers because there are too many refreshes

-> the (+) does not display anymore the room creation dialog at the home page

-> the (+) is not anymore dismissed when scrolling

-> The people section is not refreshed when a room is set/unset as a people one . (it crashes when a room is not anymore a people one).

-> The application crashes when i start a search 
java.lang.NullPointerException: Attempt to invoke virtual method 'void im.vector.adapters.ParticipantAdapterItem.displayAvatar(org.matrix.androidsdk.MXSession, android.widget.ImageView)' on a null object reference
at im.vector.adapters.PeopleAdapter$ContactViewHolder.populateViews(PeopleAdapter.java:363)
at im.vector.adapters.PeopleAdapter$ContactViewHolder.access$300(PeopleAdapter.java:343)
at im.vector.adapters.PeopleAdapter.populateViewHolder(PeopleAdapter.java:141)
at im.vector.adapters.AbsAdapter.onBindViewHolder(AbsAdapter.java:212)

-> the Matrix Contact Only switch status is not kept after rotating the device

File: vector/src/main/java/im/vector/activity/CommonActivityUtils.java [hunk 1]
```diff
@@ -575,10 +576,11 @@ public static boolean isPowerLevelEnoughForAvatarUpdate(Room aRoom, MXSession aS
...
      *
      * @param aPermissionsToBeGrantedBitMap the permissions bit map to be granted
-     * @param aCallingActivity              the calling Activity that is requesting the permissions
+     * @param aCallingActivity              the calling Activity that is requesting the permissions (or fragment parent)
+     * @param fragment                      the calling fragment that is requesting the permissions
      * @return true if the permissions are granted (synchronous flow), false otherwise (asynchronous flow)
      */
-    public static boolean checkPermissions(final int aPermissionsToBeGrantedBitMap, final Activity aCallingActivity) {
+    public static boolean checkPermissions(final int aPermissionsToBeGrantedBitMap, final Activity aCallingActivity, final Fragment fragment) {
         boolean isPermissionGranted = false;
```
Comment (sabrinaj) [inline vector/src/main/java/im/vector/activity/CommonActivityUtils.java]: Done

File: vector/src/main/java/im/vector/adapters/AbsAdapter.java [hunk 2]
```diff
@@ -0,0 +1,595 @@
...
+        return lastSection.first + lastSection.second.getNbItems() + 1;
+    }
+
+    @Override
+    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup viewGroup, int viewType) {
+        Log.i(LOG_TAG, " onCreateViewHolder for viewType:" + viewType);
+        final LayoutInflater inflater = LayoutInflater.from(viewGroup.getContext());
+        switch (viewType) {
+            case -1:
+                View headerItemView = inflater.inflate(R.layout.adapter_section_header, viewGroup, false);
```
Comment (sabrinaj) [inline vector/src/main/java/im/vector/adapters/AbsAdapter.java]: Done

File: vector/src/main/java/im/vector/adapters/AbsAdapter.java [hunk 3]
```diff
@@ -0,0 +1,595 @@
...
+                return new HeaderViewHolder(headerItemView);
+            case -2:
+                View invitationView = inflater.inflate(R.layout.adapter_item_room_invite, viewGroup, false);
```
Comment (sabrinaj) [inline vector/src/main/java/im/vector/adapters/AbsAdapter.java]: Done

File: vector/src/main/java/im/vector/adapters/AbsAdapter.java [hunk 4]
```diff
@@ -0,0 +1,595 @@
...
+                                headerViewHolder.itemView.setVisibility(View.VISIBLE);
+                            }
+                            headerViewHolder.populateViews(adapterSection.second);
+                        }
+                        break;
+                    }
+                }
+                break;
+            case -2:
+                final InvitationViewHolder invitationViewHolder = (InvitationViewHolder) viewHolder;
```
Comment (sabrinaj) [inline vector/src/main/java/im/vector/adapters/AbsAdapter.java]: Done

File: vector/src/main/java/im/vector/adapters/AbsAdapter.java [hunk 5]
```diff
@@ -0,0 +1,595 @@
...
+    }
+
+    @Override
+    public void onBindViewHolder(final RecyclerView.ViewHolder viewHolder, int position) {
+        Log.i(LOG_TAG, " onBindViewHolder for position:" + position);
+        final int viewType = getItemViewType(position);
+
+        switch (viewType) {
+            case -1:
+                final HeaderViewHolder headerViewHolder = (HeaderViewHolder) viewHolder;
```
Comment (sabrinaj) [inline vector/src/main/java/im/vector/adapters/AbsAdapter.java]: Done

File: vector/src/main/java/im/vector/adapters/AbsAdapter.java [hunk 6]
```diff
@@ -0,0 +1,595 @@
...
+
+    private static final String LOG_TAG = AbsAdapter.class.getSimpleName();
+
+    protected static final int TYPE_HEADER_DEFAULT = -1;
+
+    protected static final int TYPE_ROOM_INVITATION = -2;
+
+    private StickySectionHelper mStickySectionHelper;
+
+    private List<Pair<Integer, AdapterSection>> mSections;
```
Comment (sabrinaj) [inline vector/src/main/java/im/vector/adapters/AbsAdapter.java]: Done

File: vector/src/main/java/im/vector/adapters/AbsAdapter.java [hunk 7]
```diff
@@ -0,0 +1,595 @@
...
+     */
+    public SectionView getSectionViewForSectionIndex(int index) {
+        return mStickySectionHelper.getSectionViewForSectionIndex(index);
+    }
+
+    /**
+     * Get the item at teh given position in the adapter
+     *
+     * @param position
+     * @return
```
Comment (sabrinaj) [inline vector/src/main/java/im/vector/adapters/AbsAdapter.java]: Done

File: vector/src/main/java/im/vector/adapters/AbsAdapter.java [hunk 8]
```diff
@@ -0,0 +1,595 @@
...
+     */
+    protected void addSection(AdapterSection section) {
+        addSection(section, -1);
+    }
+
+    /**
+     * Add a section at the given index
+     *
+     * @param section
+     */
```
Comment (sabrinaj) [inline vector/src/main/java/im/vector/adapters/AbsAdapter.java]: Done

File: vector/src/main/java/im/vector/adapters/AbsAdapter.java [hunk 9]
```diff
@@ -0,0 +1,595 @@
...
+        return list;
+    }
+
+    /**
+     * Get the list of sections currently with their header position
+     *
+     * @return
+     */
+    protected List<Pair<Integer, AdapterSection>> getSectionsArray() {
+        return mSections;
```
Comment (sabrinaj) [inline vector/src/main/java/im/vector/adapters/AbsAdapter.java]: Done

File: vector/src/main/java/im/vector/adapters/PeopleAdapter.java [hunk 10]
```diff
@@ -0,0 +1,441 @@
...
+            }
+        }
+        return null;
+    }
+
+    @Override
+    protected void populateViewHolder(int viewType, RecyclerView.ViewHolder viewHolder, int position) {
+        switch (viewType) {
+            case 0:
+                // Local header
```
Comment (sabrinaj) [inline vector/src/main/java/im/vector/adapters/PeopleAdapter.java]: Done

File: vector/src/main/java/im/vector/fragments/PeopleFragment.java [hunk 11]
```diff
@@ -130,28 +224,87 @@ protected void onResetFilter() {
...
 
+            @Override
+            public void onSelectItem(ParticipantAdapterItem contact, int position) {
+                onContactSelected(contact);
+            }
+        }, this, this);
+        mRecycler.setAdapter(mAdapter);
+
+        View checkBox = mAdapter.getSectionViewForSectionIndex(2).findViewById(R.id.matrix_only_filter_checkbox);
+        if (checkBox != null && checkBox instanceof CheckBox) {
```
Comment (sabrinaj) [inline vector/src/main/java/im/vector/fragments/PeopleFragment.java]: Done

**Comment by sabrinaj** [comment]
- [ ] the (+) is not anymore dismissed when scrolling
- [ ] if we refuse the contact permission when we arrive on the screen then go accept it from settings, the local contacts section is still empty instead of being updated with local contacts data
- [ ] add javadoc to StickySectionHelper

File: vector/src/main/java/im/vector/util/StickySectionHelper.java [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+public class StickySectionHelper extends RecyclerView.OnScrollListener implements View.OnLayoutChangeListener {
+
+    private final String LOG_TAG = StickySectionHelper.class.getSimpleName();
+
+    private RecyclerView mRecyclerView;
+    private LinearLayoutManager mLayoutManager;
+
+    private List<Pair<Integer, SectionView>> mSectionViews = new ArrayList<>();
+
+    private int headerTop = 0;
```
Comment (sabrinaj) [inline vector/src/main/java/im/vector/util/StickySectionHelper.java]: Done

**Comment by ylecollen** [comment]
I also updated the mark all as read behaviour.

File: vector/src/main/java/im/vector/adapters/AbsListAdapter.java [hunk 13]
```diff
@@ -47,6 +49,32 @@ public AbsListAdapter(@LayoutRes final int layoutRes, final OnSelectItemListener
...
+
+                results.values = mFilteredItems;
+                results.count = mFilteredItems.size();
+
+                return results;
+            }
+
+            @Override
+            protected void publishResults(CharSequence constraint, FilterResults results) {
+                notifyDataSetChanged();
```
Comment (sabrinaj) [inline vector/src/main/java/im/vector/adapters/AbsListAdapter.java]: Done

File: vector/src/main/java/im/vector/adapters/AbsAdapter.java [hunk 14]
```diff
@@ -0,0 +1,595 @@
...
+     * @param pattern
+     * @return nb of items matching the filter
+     */
+    protected int filterRooms(final AdapterSection<Room> section, final String pattern) {
+        if (!TextUtils.isEmpty(pattern)) {
+            List<Room> filteredRoom = new ArrayList<>();
+            for (final Room room : section.getItems()) {
+
+                final String roomName = VectorUtils.getRoomDisplayName(mContext, mSession, room);
+                if (Pattern.compile(Pattern.quote(pattern), Pattern.CASE_INSENSITIVE)
```
Comment (sabrinaj) [inline vector/src/main/java/im/vector/adapters/AbsAdapter.java]: Done

File: vector/src/main/java/im/vector/activity/VectorHomeActivity.java [hunk 15]
```diff
@@ -1075,18 +1114,126 @@ private void processIntentUniversalLink() {
...
+                if (room.isDirectChatInvitation()) {
+                    mDirectChatInvitations.add(room);
+                } else {
+                    mRoomInvitations.add(room);
+                }
+            }
+            Collections.sort(mDirectChatInvitations, RoomUtils.getRoomsDateComparator(mSession));
+            Collections.sort(mRoomInvitations, RoomUtils.getRoomsDateComparator(mSession));
+            // the invitations are sorted from the oldest to the more recent one
+            Collections.reverse(mDirectChatInvitations);
```
Comment (sabrinaj) [inline vector/src/main/java/im/vector/activity/VectorHomeActivity.java]: Done

File: vector/src/main/java/im/vector/adapters/AbsAdapter.java [hunk 16]
```diff
@@ -0,0 +1,595 @@
...
+     * Filter the given section (of rooms) with the given pattern
+     *
+     * @param section
+     * @param pattern
+     * @return nb of items matching the filter
+     */
+    protected int filterRooms(final AdapterSection<Room> section, final String pattern) {
+        if (!TextUtils.isEmpty(pattern)) {
+            List<Room> filteredRoom = new ArrayList<>();
+            for (final Room room : section.getItems()) {
```
Comment (sabrinaj) [inline vector/src/main/java/im/vector/adapters/AbsAdapter.java]: Done

File: vector/src/main/java/im/vector/adapters/AbsAdapter.java [hunk 17]
```diff
@@ -0,0 +1,595 @@
...
+            }
+        }
+        return -1;
+    }
+
+    /**
+     * Refresh data of a section
+     */
+    public void refreshSection(final AdapterSection section) {
+        int startPos = getSectionHeaderPosition(section) + 1;
```
Comment (sabrinaj) [inline vector/src/main/java/im/vector/adapters/AbsAdapter.java]: Done

File: vector/src/main/java/im/vector/adapters/ContactAdapter.java [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+
+        /**
+         * Get the presence for the given contact
+         *
+         * @param textView
+         * @param item
+         */
+        private void loadContactPresence(final TextView textView, final ParticipantAdapterItem item) {
+            User user = null;
+            MXSession matchedSession = null;
```
Comment (sabrinaj) [inline vector/src/main/java/im/vector/adapters/ContactAdapter.java]: Done

**Pull request merged**